lib/ansible/plugins/strategy/linear.py

```diff
@@ -206,108 +204,11 @@ def run(self, iterator, play_context):
-                                        play=iterator._play,
-                                        task=new_block.get_first_parent_include(),
-                                        _hosts=self._hosts_cache,
-                                        _hosts_all=self._hosts_cache_all,
-                                    )
-                                    display.debug("filtering new block on tags")
-                                    final_block = new_block.filter_tagged_tasks(task_vars)
-                                    display.debug("done filtering new block on tags")
-
-                                included_tasks.extend(final_block.get_tasks())
-
-                                for host in hosts_left:
-                                    if host in included_file._hosts:
-                                        all_blocks[host].append(final_block)
-
-                            display.debug("done iterating over new_blocks loaded from include file")
-                        except AnsibleParserError:
-                            raise
-                        except AnsibleError as e:
-                            display.error(to_text(e), wrap_text=False)
-                            for r in included_file._results:
-                                r._result['failed'] = True
-                                r._result['reason'] = str(e)
-                                self._tqm._stats.increment('failures', r._host.name)
-                                self._tqm.send_callback('v2_runner_on_failed', r)
-                                failed_includes_hosts.add(r._host)
-                        else:
-                            # since we skip incrementing the stats when the task result is
-                            # first processed, we do so now for each host in the list
-                            for host in included_file._hosts:
-                                self._tqm._stats.increment('ok', host.name)
-                            self._tqm.send_callback('v2_playbook_on_include', included_file)
-
-                    for host in failed_includes_hosts:
-                        self._tqm._failed_hosts[host.name] = True
-                        iterator.mark_host_failed(host)
-
-                    # finally go through all of the hosts and append the
-                    # accumulated blocks to their list of tasks
-                    display.debug("extending task lists for all hosts with included blocks")
-
-                    for host in hosts_left:
-                        iterator.add_tasks(host, all_blocks[host])
-
-                    iterator.all_tasks[iterator.cur_task:iterator.cur_task] = included_tasks
-
-                    display.debug("done extending task lists")
-                    display.debug("done processing included files")
-
                 display.debug("results queue empty")
 
+                self._process_includes(results, iterator)
```

Comment:
My original intention was to call this in `StrategyBase._process_pending_results` so it's transparent to each strategy but specifically for the linear strategy this would prevent "grouping" of includes as processing includes is currently done once the results queue is empty. So to avoid any changes in behaviour I effectively just moved the common code to process `IncludedFile` from `linear` and `free` into a method in `StrategyBase` so at least the code is not duplicated.

In addition this reduces a code in a few places. Especially for `lib/ansible/playbook/included_file.py` review with `Hide whitespace`.